Adding WebView2 Unity Plugin API Reference (#2584)
MicrosoftDocs/edge-developer: microsoft-edge/webview2/reference/HoloLens2/webview2-unity-plugin.md

Reviewer: rename /HoloLens2/ to /hololens2/ lowercase per Docs Contributor Guide > Filenames
d/k if GitHub will register/reflect a mere casing change after you Refresh.  
workaround if needed: temp'ly remove '2'

Reviewer: done locally, can close if doc build is clean after i push

---

> title: WebView plugin for Unity API Reference (Preview)

Reviewer: ```suggestion
title: Mixed Reality WebView Plugin for Unity API Reference (Preview)
```
match line 11 (h1)

Reviewer: done locally, can close

description: API Reference for Microsoft Mixed Reality WebView plugin for Unity.
author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: conceptual
ms.prod: microsoft-edge
ms.technology: webview
ms.date: 05/08/2023
---

> # Mixed Reality WebView Plugin for Unity API Reference (Preview)

Reviewer: ```suggestion
# API Reference for Mixed Reality WebView Plugin (Preview)
```
check casing of "API Reference" at live: it's init-cap R:
https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/javascript/

Reviewer: done locally, can close

The **Microsoft Mixed Reality WebView plugin for Unity** enables the integration of WebView2 functionality into your HoloLens 2 app.  This WebView plugin for Unity simplifies the integration of WebView2 functionality into your HoloLens 2 app by wrapping the WebView2 control, automatically handling rendering, and automatically directing input to the WebView2 control.

This plugin also manages interop between Unity and WebView2, enabling communication between JavaScript and Unity via messages and events.

> This plugin exposes a subset of the functionality available via [CoreWebView2](/dotnet/api/microsoft.web.webview2.core.corewebview2).

Reviewer: ```suggestion
This plugin exposes a subset of the functionality that's available via [CoreWebView2](/dotnet/api/microsoft.web.webview2.core.corewebview2).
```
https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/t/that-vs-which -- "Include that even if the sentence is clear without it. It helps to clarify the sentence structure."
https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/u/using-by-using-with -- "In instructions or technical content, use by using or that use to eliminate ambiguity."

Reviewer: done locally, can close

WebView2 on HoloLens 2 and the WebView plugin for Unity are both in Preview and are subject to change before general availability.  The WebView2 Preview is available in the Insider Preview for Microsoft HoloLens.  To access this preview, you must be enrolled in the Windows Insider Program; see [Start receiving Insider builds](/hololens/hololens-insider#start-receiving-insider-builds) in _Insider preview for Microsoft HoloLens_.

WebView2 and the WebView plugin are only supported on HoloLens 2 devices running the Windows 11 update. For more information, see [Update HoloLens 2](/hololens/hololens-update-hololens).

> For information on getting started with the plugin, see [Get started with WebView2 in HoloLens 2 Unity apps](/microsoft-edge/webview2/get-started/hololens2.md).

Reviewer: ```suggestion
For information on getting started with the plugin, see [Get started with WebView2 in HoloLens 2 Unity apps](/microsoft-edge/webview2/get-started/hololens2).
```
same repo:
start with .  end with .md
different repo:
start with /  end with no .md

Reviewer: done locally... maybe keep open.  will need to merge other PR into "main" and then merge "main" into present PR to get the destination .md

Reviewer: this comment is outdated.  PR has been reduced.  
you can resolve this thread

> ### IWebView Interface

Reviewer: ```suggestion
### IWebView interface
```
Not sure what style convention to use.
Lowercase b/c:
* corp style is sentence case for titles/headings.
* C++ look & feel (if relevant) favors lowercase in API Ref docs.

> ### IWebView Interface

Reviewer: ```suggestion
### IWebView interface
```
Not sure what style convention to use.
Lowercase b/c:
* corp style is sentence case for titles/headings.
* C++ look & feel (if relevant) favors lowercase in API Ref docs.

> The `IWebView` interface provide the primary interface for the plugin.

Reviewer: ```suggestion
The `IWebView` interface is the primary interface for the plugin.
```
else "provides"

Reviewer: used "Summary" type wording:
The main interface for the WebView plugin for Unity.

done locally, can close

    Uri Page { get; }
    
    Task OnceCreated { get; }

    void Resize(int width, int height);

    void Load(Uri url);

    void Dispose();
}
```

### IWebView Delegates
```c#
public delegate void WebView_OnNavigated(string path);

public delegate void WebView_OnCanGoForwardUpdated(bool value);
```

#### IWebView Events
##### IWebView.Navigated Event
Event triggered when [CoreWebView2.SourceChanged Event](/dotnet/api/microsoft.web.webview2.core.corewebview2.SourceChanged) is raised by WebView.

###### Example
```c#
webView.Navigated += OnNavigated;

private void OnNavigated(string uri)
{
    UrlField.text = uri;
}
```

##### IWebView.WindowCloseRequested Event
Event triggered when [CoreWebView2.WindowCloseRequested Event](/dotnet/api/microsoft.web.webview2.core.corewebview2.WindowCloseRequested) is raised by WebView.

###### Example
```c#
webView.WindowCloseRequested += OnWindowCloseRequested;

private void OnWindowCloseRequested()
{
    Destroy(GameObject);
}
```

#### IWebView Properties

##### IWebView.GameObject Property
The top-level Unity GameObject entity that represents the WebView plugin in the scene. **Readonly**.

##### IWebView.Texture Property

> The 2D Unity Texture2D object that the WebView content is rendered to. **Readonly**

Reviewer: ```suggestion
The 2D Unity `Texture2D` object that the WebView content is rendered to. **Readonly**
```
global - i'm considering pushing a cleanup commit

Reviewer: done locally, can close

##### IWebView.Width Property
The width of the WebView texture and the WebView control. Note that the rendered dimensions of the IWebView instance in the Unity scene is controlled by the GameObject.

##### IWebView.Height Property

> The height of the WebView texture and the WebView control. Note that the rendered dimensions of the IWebView instance in the Unity scene is controlled by the GameObject.

Reviewer: ```suggestion
The height of the WebView texture and the WebView control. Note that the rendered dimensions of the `IWebView` instance in the Unity scene are controlled by the `GameObject`.
```
"is" might favor technical implementation.  "are" favors grammar.

Reviewer: done locally, can close

##### IWebView.Page Property
The URI currently loaded or being navigated to by the WebView control. **Readonly**

##### IWebView.OnceCreated Property
Task executes when the WebView control has been fully instantiated and ready to use. **Readonly**

###### Example
```c#
webView.OnceCreated.ContinueWith((task) => {

    // Finish seting up plugin.
    webview.Navigated += OnNavigated;
    webview.WindowCloseRequested += OnWindowCloseRequested;

    Load(initialURL);

}, TaskScheduler.FromCurrentSynchronizationContext());
```

#### IWebView Methods

##### IWebView.Resize Method

> Changes the size of the WebView2 control and the `Texture`. For additional details, see the underlying [CoreWebView2Controller.Bounds Property](/dotnet/api/microsoft.web.webview2.core.corewebview2controller.bounds).

Reviewer: ```suggestion
Changes the size of the WebView2 control and the `Texture`. For details, see the underlying [CoreWebView2Controller.Bounds Property](/dotnet/api/microsoft.web.webview2.core.corewebview2controller.bounds).
```

Reviewer: done locally, can close

> Note that the rendered dimensions of the IWebView instance in the Unity scene is controlled by the GameObject.

Reviewer: ```suggestion
Note that the rendered dimensions of the IWebView instance in the Unity scene are controlled by the GameObject.
```
"is" might favor technical implementation. "are" favors grammar.

Reviewer: done locally, can close

###### Example
```c#
Resize(600, 400);
```

##### IWebView.Load Method

> Navigates to the specificied URI. For additional details, see the underlying [CoreWebView2.Navigate Method](/dotnet/api/microsoft.web.webview2.core.corewebview2.Navigate).

Reviewer: ```suggestion
Navigates to the specified URI. For details, see the underlying [CoreWebView2.Navigate Method](/dotnet/api/microsoft.web.webview2.core.corewebview2.Navigate).
```

Reviewer: done locally, can close

###### Example

```c#
Load(new Uri("https://www.microsoft.com"));

```

##### IWebView.Dispose Method

> Clears memory, handle, callbacks, etc. related to an IWebView instance.

Reviewer: ```suggestion
Clears memory, handle, callbacks, etc. related to an `IWebView` instance.
```

Reviewer: done locally, can close

> Invoke this method when you are done with a particular IWebView instance to ensure internal memory is properly freed. Once this is called, the IWebView instance should be considered invalid.

Reviewer: ```suggestion
Invoke this method when you are done with a particular IWebView instance to ensure internal memory is properly freed. After `Dispose` is called, the `IWebView` instance should be considered invalid.
```

Reviewer: done locally

###### Example

```c#
void OnDestroy()
{
    webView.Dispose();
}

```

### IWithMouseEvents Interface

> Interface for enable mouse/pointer input for the plugin.

Reviewer: ```suggestion
Interface that supports mouse and pointer input for the plugin.
```

Reviewer: done locally, can close

> Interface for enable mouse/pointer input for the plugin.

Reviewer: ```suggestion
Supports mouse and pointer input for the plugin.
```

Reviewer: done locally, can close

```c#
public interface IWithMouseEvents
{
    void MouseEvent(WebViewMouseEventData mouseEvent);
}
```

#### IWithMouseEvents Methods

##### IWithMouseEvents.MouseEvent Method

> Propagates a `WebViewMouseEventData` event to the WebView control. Depending internal logic, the event results in the [CoreWebView2Controller.SendMouseInput Method](/dotnet/api/microsoft.web.webview2.core.corewebview2compositioncontroller.sendmouseinput) or the [CoreWebView2Controller.SendPointerInput Method](/dotnet/api/microsoft.web.webview2.core.corewebview2compositioncontroller.sendpointerinput).

Reviewer: ```suggestion
Propagates a `WebViewMouseEventData` event to the WebView control. Depending on internal logic, the event results in either the [CoreWebView2Controller.SendMouseInput Method](/dotnet/api/microsoft.web.webview2.core.corewebview2compositioncontroller.sendmouseinput) or the [CoreWebView2Controller.SendPointerInput Method](/dotnet/api/microsoft.web.webview2.core.corewebview2compositioncontroller.sendpointerinput).
```

Reviewer: done locally, can close

###### Example
```c#
public void OnPointerDown(PointerEventData eventData)
{
    IWithMouseEvents mouseEventsWebView = webView as IWithMouseEvents;

    // Call hypothetical function which converts the event's x, y into the WebView2's coordinate space.
    var hitCoord = ConvertToWebViewSpace(eventData.position.x, eventData.position.y);

    WebViewMouseEventData mouseEvent = new WebViewMouseEventData
    {
        X = hitCoord.x,
        Y = hitCoord.y,
        Type = PointerEvent.PointerDown,
        Button = PointerButton.Left,
        TertiaryAxisDeviceType = WebViewMouseEventData.TertiaryAxisDevice.PointingDevice
    };

    mouseEventsWebView.MouseEvent(mouseEvent);
}
```

### IWithPostMessage Interface
Interface for interop communication between Unity code and hosted WebView code.

> To learn more about interop in WebView2 see [Interop of native-side and web-side code](/microsoft-edge/webview2/how-to/communicate-btwn-web-native).

Reviewer: ```suggestion
To learn more about interop in WebView2, see [Interop of native-side and web-side code](../../how-to/communicate-btwn-web-native.md).
```

Reviewer: done locally, can close

```c#
public interface IWithPostMessage : IWebView
{
    event WebView_OnPostMessage MessageReceived;

    void PostMessage(string message, bool isJSON = false);
}
```

### IWithPostMessage Delegates

```c#
public delegate void WebView_OnPostMessage(string message);
```

#### IWithPostMessage Events

##### IWithPostMessage.MessageReceived Method

> Event triggered when a new JavaScript message is received from the WebView control. For additional details, see the underlying [CoreWebView2.WebMessageReceived Event](/dotnet/api/microsoft.web.webview2.core.corewebview2.webmessagereceived).

Reviewer: ```suggestion
Event triggered when a new JavaScript message is received from the WebView control. For details, see the underlying [CoreWebView2.WebMessageReceived Event](/dotnet/api/microsoft.web.webview2.core.corewebview2.webmessagereceived).
```

Reviewer: done locally, can close

> Event triggered when a new JavaScript message is received from the WebView control. For additional details, see the underlying [CoreWebView2.WebMessageReceived Event](/dotnet/api/microsoft.web.webview2.core.corewebview2.webmessagereceived).

Reviewer: ```suggestion
Event triggered when a new JavaScript message is received from the WebView control. For details, see the underlying [CoreWebView2.WebMessageReceived Event](/dotnet/api/microsoft.web.webview2.core.corewebview2.webmessagereceived).
```

###### Example
```c#
(webView as IWithPostMessage).MessageReceived += OnMessageReceived;

void OnMessageReceived(string message)
{
    Debug.Log(message);
}
```

#### IWithPostMessage Methods

##### IWithPostMessage.PostMessage Method

Sends a JavaScript message to the hosted content in the WebView control. Depending on the `isJSON` parameter, this will either result in a [CoreWebView2.PostWebMessageAsString Method](/dotnet/api/microsoft.web.webview2.core.corewebview2.postwebmessageasstring) or [CoreWebView2.PostWebMessageAsJson Method](/dotnet/api/microsoft.web.webview2.core.corewebview2.postwebmessageasjson).

```c#
var msg = new MyMessage("updateText", "Updated from Unity!");

(webView as IWithPostMessage).PostMessage(JsonUtility.ToJson(msg), true);
```

> ### IWithBrowserHistory

Reviewer: ```suggestion
### IWithBrowserHistory interface
```
not sure the doc design / style conventions.
gh rendered page: 
tbd

Reviewer: done locally, can close

Interface for handling browser-history related functionality. For example, navigating to a previous page.

```c#
public interface IWithBrowserHistory : IWebView
{
    event WebView_OnCanGoForwardUpdated CanGoForwardUpdated;

    event WebView_OnCanGoBackUpdated CanGoBackUpdated;

    void GoBack();

    void GoForward();
}
```

### IWithBrowserHistory Delegates

```c#
public delegate void WebView_OnCanGoBackUpdated(bool value);

public delegate void WebView_OnCloseRequested();
```

#### IWithBrowserHistory Events

##### IWithBrowserHistory.CanGoForwardUpdated Event

> Event triggered when a navigation occurs. The event delegate with provide a true value if [CoreWebView2.CanGoForward Property](/dotnet/api/microsoft.web.webview2.core.corewebview2.cangoforward) is true.

Reviewer: ```suggestion
Event triggered when a navigation occurs. The event delegate with provide a `true` value if [CoreWebView2.CanGoForward Property](/dotnet/api/microsoft.web.webview2.core.corewebview2.cangoforward) is `true`.
```

Reviewer: done locally, can close

###### Example
```c#
(webView as IWithBrowserHistory).CanGoBackUpdated += OnCanGoBack;

void OnCanGoBack(bool value)
{
    BackButton.enabled = value;
}
```

##### IWithBrowserHistory.CanGoBackUpdated Event

> Event triggered when a navigation occurs. The event delegate with provide a true value if [CoreWebView2.CanGoBack Property](/dotnet/api/microsoft.web.webview2.core.corewebview2.cangoback) is true.

Reviewer: ```suggestion
Event triggered when a navigation occurs. The event delegate with provide a `true` value if [CoreWebView2.CanGoBack Property](/dotnet/api/microsoft.web.webview2.core.corewebview2.cangoback) is `true`.
```

Reviewer: done locally, can close

###### Example
```c#
(webView as IWithBrowserHistory).CanGoForwardUpdated += OnCanGoForward;

void OnCanGoForward(bool value)
{
    ForwardButton.enabled = value;
}
```

#### IWithBrowserHistory Methods

##### IWithBrowserHistory.GoBack Method

> Navigates to the previous page. For additional details, see the underlying [CoreWebView2.GoBack Method](/dotnet/api/microsoft.web.webview2.core.corewebview2.goback).

Reviewer: ```suggestion
Navigates to the previous page. For details, see the underlying [CoreWebView2.GoBack Method](/dotnet/api/microsoft.web.webview2.core.corewebview2.goback).
```

Reviewer: done locally, can close

###### Example
```c#
(webView as IWithBrowserHistory).GoBack();
```

##### IWithBrowserHistory.GoForward Method

> Navigates to the next page. For additional details, see the underlying [CoreWebView2.GoForward Method](/dotnet/api/microsoft.web.webview2.core.corewebview2.goforward).

Reviewer: ```suggestion
Navigates to the next page. For details, see the underlying [CoreWebView2.GoForward Method](/dotnet/api/microsoft.web.webview2.core.corewebview2.goforward).
```

Reviewer: done locally, can close

This plugin exposes a subset of the functionality available via [CoreWebView2](/dotnet/api/microsoft.web.webview2.core.corewebview2).

WebView2 on HoloLens 2 and the WebView plugin for Unity are both in Preview and are subject to change before general availability.  The WebView2 Preview is available in the Insider Preview for Microsoft HoloLens.  To access this preview, you must be enrolled in the Windows Insider Program; see [Start receiving Insider builds](/hololens/hololens-insider#start-receiving-insider-builds) in _Insider preview for Microsoft HoloLens_.

> WebView2 is only supported on HoloLens 2 devices running the Windows 11 update. For more information, see [Update HoloLens 2](/hololens/hololens-update-hololens).

Reviewer 2: "WebView2 **and the WebView plugin** are only supported..."

What do you think?

Reviewer: latest Files Changed tab has:
WebView2 and the WebView plugin are only supported on HoloLens 2 devices running the Windows 11 update. For more information, see [Update HoloLens 2](/hololens/hololens-update-hololens).

you can close this thread

> WebView2 is only supported on HoloLens 2 devices running the Windows 11 update. For more information, see [Update HoloLens 2](/hololens/hololens-update-hololens).

Reviewer: ```suggestion
WebView2 and the WebView plugin are only supported on HoloLens 2 devices running the Windows 11 update. For more information, see [Update HoloLens 2](/hololens/hololens-update-hololens).
```
